File: Array/wuchao/string/7-18 带回退符的字符串比较/BackSpaceCompare.java
package com.slow.lib.practice.string;

/**
 * 取两个字符串的字符数组，用一个stackTopIndex 指向栈顶，在字符串数组上模拟栈的操作。
 * 遍历字符串数组，如果不是#则stackTopIndex++，并且将这个字符赋值为chars[i]的值，代表入栈
 * 如果字符为#则stackTopIndex--，代表出栈。遍历完之后则两个字符数组后，栈里面放的都是去除退格字符后的字符。
 * 最后对比两个字符数组的栈，栈容量不等返回false，所有元素都相等返回true
 * @Author wuchao
 * @Date 2020/7/18-10:47 AM
 * @description
 * @email 329187218@qq.com
 * @see
 */
public class BackSpaceCompare {
    public boolean backspaceCompare(String S, String T) {
        char[] sChars = S.toCharArray();
        char[] tChars = T.toCharArray();
        int stackTopS = -1,stackTopT = -1;
        for(int i=0;i<sChars.length;i++){
            if(sChars[i]!='#'){
                sChars[++stackTopS] = sChars[i];
            } else {
                if(stackTopS==-1) continue;
                stackTopS--;
            }
        }
        for(int i=0;i<tChars.length;i++){
            if(tChars[i]!='#'){
                tChars[++stackTopT] = tChars[i];
            } else {
                if(stackTopT==-1) continue;
                stackTopT--;
            }
        }
        if(stackTopT!=stackTopS) return false;
        if(stackTopS==-1&&stackTopT==-1) return true;
        for(int i=0;i<=stackTopS;i++){
            if(sChars[i]!=tChars[i]){
                return false;
            }
        }
        return true;
    }
}

File: Array/wuchao/string/7-18 带回退符的字符串比较/BackSpaceCompare.kt
package com.slow.lib.practice.string

/**
 * 另一种思路，从字符串尾部开始从前遍历，分别对比两个字符串的字符，如果遇到#则跳过#之前的那个字符
 * 直到遍历完字符串。中间遇到字符不相等即返回false，如果其中一个字符串遍历完而另一个没遍历完还剩不是#的字符也返回false
 * 如果跳过所有#前面的字符后，其他字符都相等，则返回true
 * @Author wuchao
 * @Date 2020/7/18-10:47 AM
 * @see
 * @description
 * @email 329187218@qq.com
 */
class BackSpaceCompare {
    fun backspaceCompare(S: String, T: String): Boolean {
        var i = S.length - 1
        var j = T.length - 1
        var skipCountS = 0
        var skipCountT = 0
        while (i >= 0 || j >= 0) {
            while (i >= 0) {
                if(S[i]=='#'){
                    skipCountS++
                    i--
                    continue
                }
                if(skipCountS>0){
                    i--
                    skipCountS--
                    continue
                }
                break
            }
            while(j>=0){
                if(T[j]=='#'){
                    skipCountT++
                    j--
                    continue
                }
                if(skipCountT>0){
                    j--
                    skipCountT--
                    continue
                }
                break
            }
            if(i>=0&&j>=0&&S[i]!=T[j]){
                return false
            }
            if((i-->=0)!=(j-->=0)){
                return false
            }

        }
        return true

    }
}